src/shogun/kernel/MaternKernel.cpp

```diff
@@ -0,0 +1,197 @@
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <shogun/distance/EuclideanDistance.h>
+#include <shogun/kernel/MaternKernel.h>
+
+using namespace shogun;
+
+MaternKernel::MaternKernel() : ShiftInvariantKernel()
+{
+	auto distance = std::make_shared<EuclideanDistance>();
+	distance->set_disable_sqrt(false);
+	m_distance = distance;
+	SG_ADD(
+	    &m_nu, "nu", "order of the modified Bessel function of the second kind",
+	    ParameterProperties::HYPER)
+	SG_ADD(
+	    &m_width, "width", "Kernel scale",
+	    ParameterProperties::HYPER | ParameterProperties::GRADIENT)
+}
+
+MaternKernel::MaternKernel(int32_t size, float64_t w, float64_t order)
+    : MaternKernel()
+{
+	m_nu = order;
+	m_width = w;
+	set_cache_size(size);
+}
+
+MaternKernel::MaternKernel(float64_t w, float64_t order)
+    : MaternKernel(10, w, order)
+{
+}
+
+MaternKernel::MaternKernel(
+    const std::shared_ptr<Features>& l, const std::shared_ptr<Features>& r,
+    int32_t size, float64_t w, float64_t order)
+    : MaternKernel(size, w, order)
+{
+	init(l, r);
+}
+
+MaternKernel::~MaternKernel()
+{
+	cleanup();
+}
+
+void MaternKernel::cleanup()
+{
+	Kernel::cleanup();
+	m_distance->cleanup();
+}
+
+bool MaternKernel::init(
+    std::shared_ptr<Features> l, std::shared_ptr<Features> r)
+{
+	cleanup();
+	return ShiftInvariantKernel::init(l, r);
+}
+
+float64_t MaternKernel::compute(int32_t idx_a, int32_t idx_b)
+{
+	float64_t result;
+	const float64_t dist = ShiftInvariantKernel::distance(idx_a, idx_b);
+
+	// first we can check if we should use one of the approximations which are
+	// cheaper to calculate these formulas were taken from Rasmussen's GPs for
+	// ML
+	if (Math::fequals(m_nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
+		result = std::exp(-dist / m_width);
+	else if (Math::fequals(
+	             m_nu, 1.5, std::numeric_limits<float64_t>::epsilon()))
+	{
+		float64_t ratio = (std::sqrt(3) * dist) / m_width;
+		result = (1 + ratio) * std::exp(-ratio);
+	}
+	else if (Math::fequals(
+	             m_nu, 2.5, std::numeric_limits<float64_t>::epsilon()))
+	{
+		float64_t ratio = (std::sqrt(5) * dist) / m_width;
+		result = (1 + ratio +
+		          ((5 * std::pow(dist, 2)) / (3 * std::pow(m_width, 2)))) *
+		         std::exp(-ratio);
+	}
+	// if none of the above was true we compute the value the expensive way
+	// in theory if nu is large we could calculate this with the squared
+	// exponential covariance function
+	else
+	{
+		const float64_t adjusted_dist =
+		    Math::fequals(dist, 0.0, std::numeric_limits<float64_t>::epsilon())
+		        ? std::numeric_limits<float32_t>::epsilon()
+		        : dist;
+		float64_t ratio = std::sqrt(2 * m_nu) * adjusted_dist / m_width;
+		// bessel function of the second kind
+#ifdef _LIBCPP_VERSION
+		error(
+		    "Shogun compiled with libc++ does not support Matern kernels with "
+		    "width other than 0.5, 1.5 and 2.5. The current width is {}.",
+		    m_width);
```

Comment:
it's too complicated to add support for cyl_bessel_k, so will just throw an error when running this part of the code when compiled with libstdc++. Note that realistically no one would ever use a value other than 0.5, 1.5 and 2.5..